### ISSUE TITLE
🔥High Priority Bug with v6 - Firestore startAfter query

### ISSUE THREAD

**Issue by danielfein** [invertase/react-native-firebase#2719]
Describe your issue here

Prior to v6, this worked perfectly, but with following the migration this error has popped up. What I'm experiencing is running the same query twice, except one time without an offset and the following time with an offset (i.e. this is for infinite scroll; first load has no offset, future times do).

The error message that I'm receiving does not align with the issue that I'm seeing which is that the index does not exist. If that were true then the first call would not run, and when clicking the link through to the Firebase console to set up the index, it would allow me to create it (instead, it tells me the index is already active, which it is).

So the error message and the bug do not align, but the startAfter is the culprit of the bug, which is hit when I run this code:

```
export async function getPosts(filter, offset){
    let authUser = await getAuthUser()
    let user = 'general'
    if(filter === 'followers'){
      user = authUser.uid
    }
    var query = firebase.firestore().collection(`feeds/${user}/posts`)
    .where('ratingOnly', '==', false)
    .orderBy('createdDate', 'desc')

    if(offset){
      query = query.startAfter(offset)
    }

    query = query.limit(10)

    return query.get().then(function(postsData) {
        var newPosts = []
        var newOffset = null
        postsData.forEach(function(doc) {
            newPosts.push({id: doc.id, ...doc.data()})
            newOffset = doc
        });
        return { newPosts, newOffset }
      })
      .catch(function(error) {
        console.log("Error getting initial home feed: ", error);
        return error
      });
}
```

The key bits to this code are that it takes the last document, calls it newOffset, and that's what is delivered back to it on the 2nd+ times run. Removing the line that adds the startAfter makes the code continuously run with the same data requests (i.e. without pagination).

This is a high priority bug since it's not easily catchable and I suspect a lot of other users will see this as they try to paginate.
---

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
platform :ios, '9.0'

target 'Tabs' do
  # Pods for Tabs
  pod 'RNFBApp', :path => '../node_modules/@react-native-firebase/app' 
  pod 'RNFBAuth', :path => '../node_modules/@react-native-firebase/auth'
  pod 'RNFBFirestore', :path => '../node_modules/@react-native-firebase/firestore'
  pod 'RNFBStorage', :path => '../node_modules/@react-native-firebase/storage'
  pod 'Fabric', '~> 1.10.2'
  pod 'Crashlytics', '~> 3.13.2'
  pod 'GoogleSignIn', '~> 5'
  pod 'React', :path => '../node_modules/react-native',  :subspecs => [ 
    'RCTImage',
  ]

  pod 'yoga', :path => '../node_modules/react-native/ReactCommon/yoga'
  pod 'TwitterKit'
  pod 'RNFS', :path => '../node_modules/react-native-fs'
  pod 'RNVectorIcons', :path => '../node_modules/react-native-vector-icons'
  pod 'BVLinearGradient', :path => '../node_modules/react-native-linear-gradient'
  pod 'react-native-blur', :path => '../node_modules/@react-native-community/blur'

end

target 'Tabs-tvOS' do
  target 'Tabs-tvOSTests' do
  end

end

post_install do |installer|
  installer.pods_project.targets.each do |target|

    # The following is needed to ensure the "archive" step works in XCode.
    # It removes React & Yoga from the Pods project, as it is already included in the main project.
    # Without this, you'd see errors when you archive like:
    # "Multiple commands produce ... libReact.a"
    # "Multiple commands produce ... libyoga.a"

    targets_to_ignore = %w(React yoga)
    
    if targets_to_ignore.include? target.name
      target.remove_from_project
    end

  end
end
```

#### `AppDelegate.m`:

```objc
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "AppDelegate.h"

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import <Firebase.h>
#import "RNNotifications.h"
#import <TwitterKit/TWTRKit.h>
#import "RNSplashScreen.h"

@implementation AppDelegate

#define TWITTER_SCHEME @"twitterkit-PtC5OrOhqaHFeZwbSktsT1uOF"
#define TWITTER_SCHEME_LOWER @"twitterkit-ptc5orohqahfezwbsktst1uof"

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [FIRApp configure];
  [RNNotifications startMonitorNotifications];
  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                                   moduleName:@"Tabs"
                                            initialProperties:nil];
  
//  [[Twitter sharedInstance] startWithConsumerKey:TWITTER_SCHEME consumerSecret:TWITTER_SECRET];
  
  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [RNSplashScreen showSplash:@"LaunchScreen" inRootView:rootViewController.view];

  return YES;
}

//Added for Notifications:
- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
  [RNNotifications didRegisterForRemoteNotificationsWithDeviceToken:deviceToken];
}

//Added for Notifications:
- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error {
  [RNNotifications didFailToRegisterForRemoteNotificationsWithError:error];
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
   NSURL *jsUrl = [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
  [[RCTBundleURLProvider sharedSettings] setJsLocation:jsUrl.host];
  return jsUrl;
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<NSString *,id> *)options {
  if ([[url scheme] isEqualToString:TWITTER_SCHEME]) {
    NSLog(@"This one doesn't work");
    return [[Twitter sharedInstance] application:app openURL:url options:options];
  }
  if ([[url scheme] isEqualToString:TWITTER_SCHEME_LOWER]) {
    NSLog(@"This one is fine");
    return [[Twitter sharedInstance] application:app openURL:url options:options];
  }
  
  return NO;
}

@end

```

</p>
</details>

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
info 
  React Native Environment Info:
    System:
      OS: macOS 10.15
      CPU: (8) x64 Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
      Memory: 21.76 MB / 8.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.16.3 - /usr/local/bin/node
      Yarn: 1.19.0 - ~/.yarn/bin/yarn
      npm: 6.9.0 - /usr/local/bin/npm
    SDKs:
      iOS SDK:
        Platforms: iOS 13.0, DriverKit 19.0, macOS 10.15, tvOS 13.0, watchOS 6.0
    IDEs:
      Xcode: 11.0/11A420a - /usr/bin/xcodebuild
    npmPackages:
      react: 16.8.3 => 16.8.3 
      react-native: ^0.59.9 => 0.59.10 
    npmGlobalPackages:
      react-native-cli: 2.0.1

```

- **Platform that you're experiencing the issue on**:
  - [x ] iOS
  - [ ] Android
  - [x ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`Firebase` module(s) you're using that has the issue:**
  - `Firestore`
- **Are you using `TypeScript`?**
  - `N`
  
</p>
</details>

**Comment by danielfein** [comment]
I tried the same code on web, and it worked well. I think that narrows it down to a react-native-firebase package issue. Any ideas?

The code I've run on web, yielding the correct result:

  ```
 var getHomeFeedQuery = this.db.collection(`feeds/${uid}/posts`)
    .where('ratingOnly', '==', false)
    .orderBy('createdDate', 'desc')
    if(offset){
      getHomeFeedQuery = getHomeFeedQuery.startAfter(offset)
    }
```

**Comment by danielfein** [comment]
I am running this with a query that uses `array-contains` instead of `==` and it seems it may be working on that edge case. `where('ratingOnly', '==', false) is not.

**Comment by Ehesp** [comment]
Hello, I'm on my phone so may not see it. What rnfb version are you using? 6.0.1 had some updates with query mutation and logging index creation errors so maybe worth upgrading/downgrading.

**Comment by danielfein** [comment]
Yep I'm on 6.0.1. I'm going to downgrade to 6.0.0 to see if there's a difference!

**Comment by danielfein** [comment]
The issue exists on both 6.0.0 and 6.0.1.

**Comment by Ehesp** [comment]
Thanks. Could you post an example array with a few documents/json (so I can test against). Once I'm on my computer later I'll take a look.

**Comment by Ehesp** [comment]
Also, if possible could you create a basic test case. That will help get to the bottom of the issue quicker. For example: https://github.com/invertase/react-native-firebase/blob/master/packages/firestore/e2e/Query/startAfter.e2e.js#L150

Post it here if you want, or PR it and the tests will run on the CI automatically.

**Comment by danielfein** [comment]
Yeah definitely can share the sample documents. I haven't really worked with testing so I wouldn't trust myself to write a case.

The sample data looks like this:

``` [
  {
    name: "The Avengers",
    createdDate: new Date(),
    ratingOnly: true,
    rating: 3.6
  },{
    name: "Joker",
    createdDate: new Date(),
    ratingOnly: false,
    rating: 4.5
  },{
    name: "The Avengers 2",
    createdDate: new Date(),
    ratingOnly: true,
    rating: 3
  }
}]```

The call is looking to pull from this where ratingOnly is false and in order of createdDate descending. It was working in previous versions and for some reason seems to be working for startAfter queries where the operator is array-contains. Thanks a ton! Will try to take a stab at the test case, but I wouldn't count on it working!

**Comment by danielfein** [comment]
Can you try this?


  it('startAfter snapshot', async () => {
    const colRef = firebase.firestore().collection('v6/endsAt/snapshot');
    const doc1 = colRef.doc('doc1');
    const doc2 = colRef.doc('doc2');
    const doc3 = colRef.doc('doc3');

    await Promise.all([doc1.set({ foo: 1, bar: 1, array: ['blah'] }), doc2.set({ foo: 2, bar: 2, array: ['blah'] }), doc3.set({ foo: 1, bar: 3, array: ['blah'] })]);

    const startAfter = await doc2.get();

    const qs = await colRef.where('foo', '==', 1).startAfter(startAfter).get();
    const aqs = await colRef.where('array', '==', blah).startAfter(startAfter).get();

    qs.docs.length.should.eql(1);
    qs.docs[0].id.should.eql('doc3');
 
    // this seems to be the issue:
    let data = qs.data()
    data.bar.should.eql(3);

   // this seems to be working:
    let arrayData = aqs.data()
    arrayData.array[0].should.eql('blah');
  });

**Comment by Ehesp** [comment]
Thanks will take a look. Firestore had a big rewrite on the query side of things. Infact, the start* end* methods had some underlying issues which meant they didn't work the same as the Web SDK. Part of that rewrite probably caused some other bugs (yay :D)

**Comment by danielfein** [comment]
yaaaay. let me know how I can help, I attempted a PR, hopefully it makes sense

**Comment by ogtfaber** [comment]
I have got the same issue with a paginated query.

**Comment by danielfein** [comment]
@Ehesp do you know when this issue may be looked at? Would love to have a potential timeline if possible, I do think startAfter is a critical feature to most applications. Thanks again!

**Comment by mikehardy** [comment]
I crosslinked the PR and left you a note over there - first step is to get that test PR to successfully run - which is to say fail in the expected way based on your report and thus demonstrate the issue

**Comment by danielfein** [comment]
Cool thanks a lot! Hopefully someone can pick it up soon :)

**Comment by Ehesp** [comment]
Sorry have been super busy these last few days. I should have some time tomorrow to take a look.

**Comment by Ehesp** [comment]
@danielfein Ok so I'm still a bit confused as to what the issue is here. 

Your test case is doing `.where('array', '==', blah)` - what is `blah` as it's undeclared? Generally the startAfter tests are passing, and do successfully start after documents so I'm just trying to fully understand where the issue occurs.

**Comment by danielfein** [comment]
I have a list of documents in a sub-collection located at `feeds -> userId -> posts`. The code I have is the getPosts code in the initial post above where offset is set as the last doc to run through the forEach loop from the returned data set.

The code I shared for the e2e is untested and I was writing it without testing it as I'm new to it/didn't know how. I would basically ignore the test at this point. It's the getPosts test where I'm calling for something that requires an index to be made, once that index is made it's still claiming I need an index.

This is the key bit:

    ```
var query = firebase.firestore().collection(`feeds/${user}/posts`)
    .where('ratingOnly', '==', false)
    .orderBy('createdDate', 'desc')

    if(offset){
      query = query.startAfter(offset)
    }

    query = query.limit(10)
```

It just broke in 6.0.0 so I know the code should be working, it also works in web. The only data relevant here is to have documents in a subcollection called posts with "ratingOnly" of boolean values and "createdDate" as timestamps. Hopefully this helps. Are you in a slack or anything I can try to chat with you in? Maybe the firebase community slack?

**Comment by danielfein** [comment]
I believe I've narrowed it down further to `orderBy` in conjunction with startAfter being the issue.

**Comment by ThomasFindlay** [comment]
Also got the same issue. I am getting an error about firestore/failed-precondition when I use startAfter. My situation and code is the same as @danielfein. Works when I remove startAt, but with it, it fails. Worked with v5, stopped in v6.

**Comment by danielfein** [comment]
@thomasfindlay are you using orderBy when you hit the error as well?

**Comment by ThomasFindlay** [comment]
@danielfein Yes, I am. orderBy('metadata.created', 'desc')

**Comment by Ehesp** [comment]
Hi All

Would you be able to try the following:

1. Locally open the `firestore/lib/FirestoreQuery.js` file in your Node Modules.
2. Apply the following changes to that file: https://github.com/invertase/react-native-firebase/commit/f520406520dfe80b057fb35579255d21935cf19a
3. Reload (may have to rebuild?) the app/code you're having issues with.

I've tested locally and it seems to be working. 

In a nutshell; when you apply an `orderBy` with a cursor query (`startAfter`) we were applying the order value AND document ID to the query. This change only applies the document ID if no cursor query is provided.

**Comment by ThomasFindlay** [comment]
@Ehesp Just applied the change you asked. Using startAfter does not result in an error.

**Comment by Ehesp** [comment]
@ThomasFindlay great! Does it result in the expected data? I've tried with a limited data set so would be good to test this out on a large list.

**Comment by ThomasFindlay** [comment]
@Ehesp unfortunately, I do not have a large list of data to test it on at the moment.

**Comment by Salakar** [comment]
v6.0.2 release is now published with this fix in - thanks
